src/Hazelcast.Net/Core/ReadOptimizedLruCache.cs

```diff
@@ -0,0 +1,165 @@
+        if (capacity >= threshold) throw new ArgumentException("Threshold value cannot be less than or equal capacity.");
+
+        _capacity = capacity;
+        _threshold = threshold;
+    }
+
+
+    /// <summary>
+    /// Try get value corresponds to given key.
+    /// </summary>
+    /// <param name="key">Key</param>
+    /// <param name="val">Value</param>
+    /// <returns>True if key exists or false.</returns>
+    /// <exception cref="ArgumentNullException">If key is null.</exception>
+    public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue val)
+    {
+        if (_disposed == 1) throw new ObjectDisposedException("Cache is disposed.");
+
+        if (key is null) throw new ArgumentNullException(nameof(key));
+
+        if (Cache.TryGetValue(key, out var entry))
+        {
+            entry.Touch();
+            val = entry.Value;
+            return true;
+        }
+
+        val = default;
+        return false;
+    }
+
+    /// <summary>
+    /// Adds the given key value to cache.
+    /// </summary>
+    /// <param name="key">Key</param>
+    /// <param name="val">Value</param>
+    /// <exception cref="ArgumentNullException">If key is null.</exception>
+    public void Add(TKey key, TValue val)
+    {
+        if (_disposed == 1) throw new ObjectDisposedException("Cache is disposed.");
+
+        if (key is null) throw new ArgumentNullException(nameof(key));
+
+        var entry = new TimeBasedEntry<TValue>(val);
+        var addVal = Cache.GetOrAdd(key, entry);
+
+        if (!addVal.Equals(val) && Cache.Count > _threshold)
+            DoEviction();
+    }
+
+    /// <summary>
+    /// Try to remove key value pair by given key.
+    /// </summary>
+    /// <param name="key">Key to be removed</param>
+    /// <param name="value">Value is removed, default value if key not exists.</param>
+    /// <returns>True if key pair removed, otherwise false.</returns>
+    /// <exception cref="ObjectDisposedException">If cache disposed</exception>
+    /// <exception cref="ArgumentNullException">If key null.</exception>
+    public bool TryRemove(TKey key, [MaybeNullWhen(false)] out TValue value)
+    {
+        if (_disposed == 1) throw new ObjectDisposedException("Cache is disposed.");
+
+        if (key is null) throw new ArgumentNullException(nameof(key));
+
+        if (Cache.TryRemove(key, out var entry))
+        {
+            value = entry.Value;
+            return true;
+        }
+
+        value = default;
+        return false;
+    }
+
+    private void DoEviction()
+    {
+        // Don't block if a thread already doing eviction.
+        if (_cleaningSlim.CurrentCount == 0) return;
+
+        _cleaningSlim.Wait();
+
+        try
+        {
+            if (Cache.Count < _threshold) return;
+
+            var countOfEntriesToRemoved = Cache.Count - _capacity;
+#if NET6_0_OR_GREATER
```

Comment:
have you benchmarked these methods? on my machine,
```var cutoff = Cache.OrderBy(x => x.Value.LastTouch).ElementAt(Index).Value.LastTouch;```
is the fastest option, putting the `long` values in a `List<long>` and then using `List.Sort()` is only slightly slower, and the priority queue method is generally slower. Or, almost all of the time. What do you think?

Reply:
I've committed a benchmark min-heap vs ordering at `Hazelcast.Net.Benchmarks/LruCacheEviction.cs` . Min-heap performed better in any size of cache. Also, Java implementation has similar results here https://github.com/hazelcast/hazelcast/pull/22659#discussion_r1111783551 . I'm surprised that you found the opposite.

Reply:
strange indeed. that being said, my benchmark just determines the cutoff value, it does not do the whole eviction. ok, let's say it's fine the way it is now.